GH-41134: [GLib] Support building arrow-glib with MSVC

adamreeve opened the pull request:
### Rationale for this change

Allow Windows users to more easily build the GLib libraries.

### What changes are included in this PR?

* Minor fixes to allow building with MSVC:
  * Changes some uses of variable length arrays to `std::vector`, because MSVC doesn't support variable length arrays
  * Moves some function definitions that use C++ types outside of `G_BEGIN_DECLS`/`G_END_DECLS` blocks (which expand to `extern C { }`), because this caused MSVC to error
* Fix libraries not having any exported symbols with MSVC, which defaults to hiding symbols
  * Add `visibility.h` which defines a new `GARROW_EXTERN` macro that adds `dllimport` or `dllexport` attributes when using MSVC.
  * Include the `GARROW_EXTERN` macro in the definitions of the `GARROW_AVAILABLE_IN_*` macros.
* Add a new CI job that builds the GLib libraries with MSVC on Windows, using vcpkg to install pkgconfig and glib.
* For now only `arrow-glib` is built, I can follow up with the other libraries after this PR. That will require introducing new per-library version macros.

### Are these changes tested?

The build will be tested in CI but I've only done some quick manual tests that the built library works correctly, I haven't got the ruby tests running against the build yet.

### Are there any user-facing changes?

No? Eventually some documentation should be updated when all the GLib libraries can be built with MSVC though
* GitHub Issue: #41134

kou commented on .github/workflows/ruby.yml:
```diff
@@ -305,3 +305,96 @@ jobs:
+        # so we need to make sure the MSCV link.exe is first in $PATH
+        run: |
+          call "C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Auxiliary\Build\vcvarsall.bat" x64
+          set CMAKE_BUILD_PARALLEL_LEVEL=%NUMBER_OF_PROCESSORS%
```
Do we need this?
It seems that we use Ninja and Ninja uses parallel build by default.

adamreeve replied:
Yeah we shouldn't need this. I copied this from the MinGW windows build steps above but I'm guessing it isn't used due to using Meson rather than CMake anyways. I'll remove it from there too

kou commented on c_glib/arrow-glib/array-builder.cpp:
```diff
@@ -231,8 +231,14 @@ garrow_array_builder_append_values(GArrowArrayBuilder *builder,
     ++n_loops;
   }
+#ifdef _MSVC_LANG
+  std::vector<uint8_t> data_vec(value_size * chunk_size);
+  uint8_t *data = data_vec.data();
```
We can always use this style. (We can remove `uint8_t data[value_size * chunk_size]` style.)

adamreeve replied:
My understanding is that the variable length array approach would use stack allocation so there might be a performance benefit over using `std::vector`, but I'm happy to only use `std::vector` if you're not worried about that.

kou replied:
Right. I want to choose easy to maintain for now. But we may rethink about this later.

kou commented on c_glib/arrow-glib/array-builder.h:
```diff
@@ -22,70 +22,97 @@
 garrow_array_builder_finish(GArrowArrayBuilder *builder, GError **error);
 
 GARROW_AVAILABLE_IN_2_0
+GARROW_EXPORT
```
Let's add this macro to `GARROW_AVAILABLE_IN_XXX`. GLib does it.

kou replied:
Right. We need separated version macros per library.

adamreeve replied:
I thought I could have this PR only change the arrow-glib library and do the others in a follow up to keep the changes manageable, but the MinGW build fails with link errors like "redeclared without dllimport attribute after being referenced with dll linkage", so I might need to add the separated version macros as part of this PR too.

adamreeve replied:
I've changed the macro definitions to only add dllimport/dllexport for MSVC for now, so this works as currently only the arrow-glib library is built for MSVC. But I can follow up to also add them under MinGW GCC later after adding version macros per-library. Does that work for you or would you rather include the per-library macros in this PR?

adamreeve replied:
I like that idea, I've made #41681 for this and will make a PR for that before updating this PR.

kou commented on ci/scripts/install_vcpkg.sh:
```diff
@@ -25,7 +25,7 @@ if [ "$#" -lt 1 ]; then
 
 arrow_dir=$(cd -- "$(dirname -- "$0")/../.." && pwd -P)
-default_vcpkg_version=$(cat "${arrow_dir}/.env" | grep "VCPKG" | cut -d "=" -f2 | tr -d '"')
+default_vcpkg_version=$(cat "${arrow_dir}/.env" | grep "VCPKG" | cut -d "=" -f2 | cut -d "#" -f1 | tr -d '" ')
```
```suggestion
default_vcpkg_version=$(source "${arrow_dir}/.env" | echo "$VCPKG")
```

kou commented on ci/scripts/c_glib_build.sh:
```diff
@@ -35,8 +36,16 @@ export CXXFLAGS="-DARROW_NO_DEPRECATED_API"
 
 mkdir -p ${build_dir}
 
+if [ "${BUILD_C_GLIB_WITH_VCPKG}" == "ON" ]; then
```
Can we use existing `${VCPKG_ROOT}` instead of new variable?

```suggestion
if [ -n "${VCPKG_ROOT:-}" ]; then
```

adamreeve replied:
Yeah, I thought there might be cases where VCPKG_ROOT is defined but we don't necessarily want to use vcpkg, but I guess that's not very likely to happen.

kou commented on ci/scripts/c_glib_build.sh:
```diff
@@ -35,8 +36,16 @@ export CXXFLAGS="-DARROW_NO_DEPRECATED_API"
+if [ "${BUILD_C_GLIB_WITH_VCPKG}" == "ON" ]; then
+    vcpkg_install_root="${build_root}/vcpkg_installed"
+    $VCPKG_ROOT/vcpkg install --x-manifest-root=${source_dir} --x-install-root=${vcpkg_install_root}
+    export PKG_CONFIG="${vcpkg_install_root}/x64-windows/tools/pkgconf/pkgconf.exe"
+    export PKG_CONFIG_PATH="${vcpkg_install_root}/x64-windows/lib/pkgconfig:${PKG_CONFIG_PATH}"
```
Could you use `meson setup`'s `--pkg-config-path` instead?

kou commented on c_glib/vcpkg.json:
```diff
@@ -0,0 +1,8 @@
+{
+  "name": "arrow-glib",
+  "version-string": "",
```
```suggestion
  "version-string": "17.0.0-SNAPSHOT",
```

We can update this automatically in our release process:

```diff
diff --git a/dev/release/01-prepare-test.rb b/dev/release/01-prepare-test.rb
index 8fb23f45f0..11e7561281 100644
--- a/dev/release/01-prepare-test.rb
+++ b/dev/release/01-prepare-test.rb
@@ -108,6 +108,13 @@ class PrepareTest < Test::Unit::TestCase
            "+version = '#{@release_version}'"],
         ],
       },
+      {
+        path: "c_glib/vcpkg.json",
+        hunks: [
+          ["-  \"version-string\": \"#{@snapshot_version}\",",
+           "+  \"version-string\": \"#{@release_version}\","],
+        ],
+      },
       {
         path: "ci/scripts/PKGBUILD",
         hunks: [
diff --git a/dev/release/post-11-bump-versions-test.rb b/dev/release/post-11-bump-versions-test.rb
index 78d9320bfb..818a0a4b61 100644
--- a/dev/release/post-11-bump-versions-test.rb
+++ b/dev/release/post-11-bump-versions-test.rb
@@ -86,6 +86,13 @@ class PostBumpVersionsTest < Test::Unit::TestCase
            "+version = '#{@next_snapshot_version}'"],
         ],
       },
+      {
+        path: "c_glib/vcpkg.json",
+        hunks: [
+          ["-  \"version-string\": \"#{@snapshot_version}\",",
+           "+  \"version-string\": \"#{@next_snapshot_version}\","],
+        ],
+      },
       {
         path: "ci/scripts/PKGBUILD",
         hunks: [
diff --git a/dev/release/utils-prepare.sh b/dev/release/utils-prepare.sh
index 5136708722..3de33625d3 100644
--- a/dev/release/utils-prepare.sh
+++ b/dev/release/utils-prepare.sh
@@ -40,6 +40,12 @@ update_versions() {
     meson.build
   rm -f meson.build.bak
   git add meson.build
+
+  sed -i.bak -E -e \
+    "s/\"version-string\": \".+\"/\"version-string\": \"${version}\"/" \
+    vcpkg.json
+  rm -f vcpkg.json.bak
+  git add vcpkg.json
   popd
 
   pushd "${ARROW_DIR}/ci/scripts"
```

kou reviewed (commented):
Thanks!

adamreeve commented on c_glib/arrow-glib/visibility.h:
```diff
@@ -0,0 +1,56 @@
+#pragma once
+
+#if defined(_WIN32) || defined(__CYGWIN__)
+// Windows
```
:+1: This was mostly copied from `cpp/src/arrow/util/visibility.h` and simplified by the way

adamreeve commented on .github/workflows/ruby.yml:
```diff
@@ -305,3 +304,96 @@ jobs:
+      ARROW_FLIGHT: OFF
+      ARROW_FLIGHT_SQL: OFF
+      ARROW_HDFS: OFF
+      ARROW_HOME: "${{ github.workspace }}/dist"
```
I had to change `ARROW_HOME` to be in the workspace directory rather than `/usr`, as `/usr` becomes `C:\Program Files\Git\usr` as a Windows path and I was getting errors when linking like:
```
[47/54] Linking target arrow-glib/arrow-glib-1700.dll
FAILED: arrow-glib/arrow-glib-1700.dll arrow-glib/arrow-glib-1700.pdb 
"link" @arrow-glib/arrow-glib-1700.dll.rsp
LINK : fatal error LNK1181: cannot open input file 'Files\Git\usr\bin.obj'
```

I didn't get to the bottom of where this was going wrong but it seems like something in meson is failing to quote the linker command inputs.

adamreeve commented on .github/workflows/ruby.yml:
```diff
@@ -305,3 +305,96 @@ jobs:
+        env:
+          # We can invalidate the current cache by updating this.
+          CACHE_VERSION: "2024-05-09"
+      - name: Build
```
I looked into this and got it working on another branch (https://github.com/adamreeve/arrow/tree/msvc_glib_vcpkg), but it's quite slow to build all the vcpkg dependencies (48 minutes on my GitHub actions run) and this isn't integrated with ccache so will be slow every time. If you think it's worth doing I can look at caching the vcpkg builds too? We do this in ParquetSharp so it shouldn't be too hard to add: https://github.com/G-Research/ParquetSharp/blob/ef8e7c447ecd5d84205eb73293162f5689e977b8/.github/workflows/ci.yml#L99-L118

kou replied:
OK. Let's work on it as a separated task.

FYI: We have a job that uses `VCPKG_BINARY_SOURCES`: https://github.com/apache/arrow/blob/657c4faf21700c0899703a4759bde76235c38199/dev/tasks/vcpkg-tests/github.windows.yml#L27
It may also work.

raulcd commented:
@github-actions crossbow submit -g release

raulcd commented:
@github-actions crossbow submit -g nightly-release

kou commented on c_glib/arrow-glib/buffer.h:
```diff
@@ -20,58 +20,90 @@
 #pragma once
 
 #include <glib-object.h>
+#include <arrow-glib/version.h>
```
```suggestion

#include <arrow-glib/version.h>
```

kou commented on c_glib/arrow-glib/chunked-array.hpp:
```diff
@@ -22,11 +22,17 @@
 #include <arrow/api.h>
 
 #include <arrow-glib/chunked-array.h>
+#include <arrow-glib/visibility.h>
```
I think that this is needless.

```suggestion
```

kou commented on c_glib/arrow-glib/codec.hpp:
```diff
@@ -22,13 +22,20 @@
 #include <arrow/util/compression.h>
 
 #include <arrow-glib/codec.h>
+#include <arrow-glib/visibility.h>
```
I think that this is needless.

```suggestion
```

kou commented on c_glib/arrow-glib/compute-definition.h:
```diff
@@ -20,10 +20,12 @@
 #pragma once
 
 #include <glib-object.h>
+#include <arrow-glib/version.h>
```
```suggestion

#include <arrow-glib/version.h>
```

kou commented on c_glib/arrow-glib/error.h:
```diff
@@ -20,6 +20,7 @@
 #pragma once
 
 #include <glib-object.h>
+#include <arrow-glib/version.h>
```
```suggestion

#include <arrow-glib/version.h>
```

kou commented on c_glib/arrow-glib/error.hpp:
```diff
@@ -22,19 +22,28 @@
 #include <arrow/api.h>
 
 #include <arrow-glib/error.h>
+#include <arrow-glib/visibility.h>
```
I think that this is needless.

```suggestion
```

kou commented on c_glib/arrow-glib/expression.hpp:
```diff
@@ -22,8 +22,12 @@
 #include <arrow/compute/expression.h>
 
 #include <arrow-glib/expression.h>
+#include <arrow-glib/visibility.h>
```
I think that this is needless.

```suggestion
```

kou commented on c_glib/arrow-glib/file-system.hpp:
```diff
@@ -22,29 +22,37 @@
 #include <arrow/filesystem/api.h>
 
 #include <arrow-glib/file-system.h>
+#include <arrow-glib/visibility.h>
```
I think that this is needless.

```suggestion
```

kou commented on c_glib/arrow-glib/input-stream.hpp:
```diff
@@ -25,35 +25,50 @@
 #include <arrow/io/memory.h>
 
 #include <arrow-glib/input-stream.h>
+#include <arrow-glib/visibility.h>
```
I think that this is needless.

```suggestion
```

kou commented on c_glib/arrow-glib/memory-pool.h:
```diff
@@ -20,23 +20,32 @@
 #pragma once
 
 #include <glib-object.h>
+#include <arrow-glib/version.h>
```
```suggestion

#include <arrow-glib/version.h>
```

kou commented on c_glib/arrow-glib/output-stream.hpp:
```diff
@@ -24,24 +24,34 @@
 #include <arrow/io/memory.h>
 
 #include <arrow-glib/output-stream.h>
+#include <arrow-glib/visibility.h>
```
I think that this is needless.

```suggestion
```

kou commented on c_glib/arrow-glib/reader.hpp:
```diff
@@ -26,43 +26,63 @@
 #include <arrow/json/api.h>
 
 #include <arrow-glib/reader.h>
+#include <arrow-glib/visibility.h>
```
I think that this is needless.

```suggestion
```

kou commented on c_glib/arrow-glib/record-batch.hpp:
```diff
@@ -22,14 +22,20 @@
 #include <arrow/api.h>
 
 #include <arrow-glib/record-batch.h>
+#include <arrow-glib/visibility.h>
```
I think that this is needless.

```suggestion
```

kou commented on c_glib/arrow-glib/schema.hpp:
```diff
@@ -22,8 +22,12 @@
 #include <arrow/api.h>
 
 #include <arrow-glib/schema.h>
+#include <arrow-glib/visibility.h>
```
I think that this is needless.

```suggestion
```

kou commented on c_glib/arrow-glib/table.hpp:
```diff
@@ -23,11 +23,16 @@
 #include <arrow/ipc/api.h>
 
 #include <arrow-glib/table.h>
+#include <arrow-glib/visibility.h>
```
I think that this is needless.

```suggestion
```

kou commented on c_glib/arrow-glib/basic-array-definition.h:
```diff
@@ -20,17 +20,20 @@
 #pragma once
 
 #include <glib-object.h>
+#include <arrow-glib/version.h>
```
```suggestion

#include <arrow-glib/version.h>
```

kou commented on c_glib/arrow-glib/basic-array.hpp:
```diff
@@ -22,23 +22,34 @@
 #include <arrow/api.h>
 
 #include <arrow-glib/basic-array.h>
+#include <arrow-glib/visibility.h>
```
I think that this is needless.

```suggestion
```

kou commented on ci/scripts/c_glib_build.sh:
```diff
@@ -28,17 +28,38 @@ build_root=${2}
-export PKG_CONFIG_PATH=${ARROW_HOME}/lib/pkgconfig
+
+if [ -n "${MSYSTEM:-}" ]; then
+    # Fix ARROW_HOME when running under MSYS2
+    export ARROW_HOME="$(cygpath --unix "${ARROW_HOME}")"
```
```suggestion
  # Fix ARROW_HOME when running under MSYS2
  export ARROW_HOME="$(cygpath --unix "${ARROW_HOME}")"
```

kou commented on ci/scripts/c_glib_build.sh:
```diff
@@ -28,17 +28,38 @@ build_root=${2}
 mkdir -p ${build_dir}
 
+if [ -n "${VCPKG_ROOT:-}" ]; then
+    vcpkg_install_root="${build_root}/vcpkg_installed"
+    $VCPKG_ROOT/vcpkg install --x-manifest-root=${source_dir} --x-install-root=${vcpkg_install_root}
+    export PKG_CONFIG="${vcpkg_install_root}/x64-windows/tools/pkgconf/pkgconf.exe"
+    meson_pkg_config_path="${vcpkg_install_root}/x64-windows/lib/pkgconfig:${meson_pkg_config_path}"
```
```suggestion
  vcpkg_install_root="${build_root}/vcpkg_installed"
  $VCPKG_ROOT/vcpkg install --x-manifest-root=${source_dir} --x-install-root=${vcpkg_install_root}
  export PKG_CONFIG="${vcpkg_install_root}/x64-windows/tools/pkgconf/pkgconf.exe"
  meson_pkg_config_path="${vcpkg_install_root}/x64-windows/lib/pkgconfig:${meson_pkg_config_path}"
```

kou commented on ci/scripts/c_glib_build.sh:
```diff
@@ -28,17 +28,38 @@ build_root=${2}
+    export ARROW_HOME="$(cygpath --unix "${ARROW_HOME}")"
+fi
+
+meson_pkg_config_path="${ARROW_HOME}/lib/pkgconfig:${ARROW_HOME}/bin/pkgconfig"
```
Do we need `:${ARROW_HOME}/bin/pkgconfig`?

adamreeve replied:
Yes, when I test the build in Windows the pkgconfig files are installed to the bin directory, and without adding that I get an error:
```
Run-time dependency arrow found: NO (tried pkgconfig and cmake)

c_glib\meson.build:108:10: ERROR: Dependency "arrow" not found, tried pkgconfig and cmake
```

kou replied:
If we remove https://github.com/apache/arrow/pull/41599/files#r1602805729 , we can remove this path too.

adamreeve replied:
Yes you're right, I copied that from the C++ build (`.github/workflows/cpp.yml`) but had forgotten about it sorry. I'm not sure why the C++ build is configured like that. I've removed setting the LIBDIR variable and the extra pkgconfig path now.

kou commented on ci/scripts/c_glib_build.sh:
```diff
@@ -28,17 +28,38 @@ build_root=${2}
+fi
+
+if [ -n "${VCToolsInstallDir:-}" -a -n "${MSYSTEM:-}" ]; then
+    # Meson finds the gnu link.exe instead of MSVC link.exe when running in MSYS2/git bash,
+    # so we need to make sure the MSCV link.exe is first in $PATH
+    export PATH="$(cygpath --unix "${VCToolsInstallDir}")/bin/HostX64/x64:${PATH}"
```
```suggestion
  # Meson finds the gnu link.exe instead of MSVC link.exe when running in MSYS2/git bash,
  # so we need to make sure the MSCV link.exe is first in $PATH
  export PATH="$(cygpath --unix "${VCToolsInstallDir}")/bin/HostX64/x64:${PATH}"
```

kou reviewed (approved):
+1

kou commented on .github/workflows/ruby.yml:
```diff
@@ -305,3 +304,96 @@ jobs:
+      BOOST_SOURCE: BUNDLED
+      CMAKE_CXX_STANDARD: "17"
+      CMAKE_GENERATOR: Ninja
+      CMAKE_INSTALL_LIBDIR: bin
```
Could you remove this?

```suggestion
```

adamreeve commented:
I've rebased on top of the changes from #41721 to add per-library version headers, and added per-library visibility headers too. I've had to keep the Gandiva build disabled due to #36958 and also disabled the Flight libraries due to a missing zlib dependency, but I'm hoping that should be fixed by switching to vcpkg dependencies for the C++ library build.

kou commented on c_glib/tool/generate-visibility-header.py:
```diff
@@ -0,0 +1,68 @@
+
+
+def write_header(output_file: TextIOBase, library: str):
+    output_file.write(f"""#pragma once
```
Could you also generate license header?
Or how about adding this to `version.h` instead of creating a separated `visibility.h`?

adamreeve replied:
Yeah adding this to `version.h` makes sense, there isn't any good reason it needs to be separate.

kou commented on ci/scripts/install_vcpkg.sh:
```diff
@@ -25,7 +25,7 @@ if [ "$#" -lt 1 ]; then
 
 arrow_dir=$(cd -- "$(dirname -- "$0")/../.." && pwd -P)
-default_vcpkg_version=$(cat "${arrow_dir}/.env" | grep "VCPKG" | cut -d "=" -f2 | tr -d '"')
+default_vcpkg_version=$(source "${arrow_dir}/.env" && echo "$VCPKG" || echo "")
```
Can we remove `|| echo ""` ?

```suggestion
default_vcpkg_version=$(source "${arrow_dir}/.env" && echo "$VCPKG")
```

adamreeve replied:
This was needed because there are some tasks that run in Docker where the `.env` file isn't included (it's not explicitly included in `.dockerignore`), eg: https://github.com/apache/arrow/blob/fb61e9f7a821dcb3c753fa6d6c36eec3714c257b/ci/docker/python-wheel-manylinux.dockerfile#L72

From what I can see these explicitly set the vcpkg version so `default_vcpkg_version` isn't used. Maybe it would make more sense to only try to read the vcpkg_version from `.env` if `$2` isn't set?

kou replied:
I see. 

>Maybe it would make more sense to only try to read the vcpkg_version from `.env` if `$2` isn't set?

+1

kou reviewed (approved):
+1
